### PR TITLE
Update internal Typescript dependency from 3.9.0 to 4.3.4

### PR DESCRIPTION
**Pull request by apexskier**
This updates the project's TypeScript version from version 3.9.0 to 4.3.4.

I'm working on this in the context of https://github.com/apexskier/nova-typescript/issues/273, because I need to update types to properly type the `skipDestructiveCodeActions` arg.

There were fewer changes than I'd expect, here are the highlights:

* Several types in the tsserver protocol are now `string | tsp.SymbolDisplayPart[]` instead of `string`. I was originally playing with improving the formatting here, but decided not to include that in the scope of this change. (although I'd like to make some improvements separately)
* There are several new command types, so I've updated `server/src/tsp-command-types.ts`

Along with automated tests, I've been manually testing this underneath https://github.com/apexskier/nova-typescript and it seems to be working the same as before.

### CODE REVIEW

File: server/src/utils.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@
  */
 
 export class Deferred<T> {
-    resolve: (value?: T) => void;
+    resolve: (value?: T | PromiseLike<T>) => void;
```
Comment (rchl) [inline server/src/utils.ts:9]: In the updated Promise types the value is no longer optional so this should just be:
```suggestion
    resolve: (value: T) => void;
```

File: server/src/tsp-client.ts [hunk 2]
```diff
@@ -10,10 +10,9 @@ import * as path from 'path';
 import * as cp from 'child_process';
 import * as readline from 'readline';
 import * as decoder from 'string_decoder';
-import protocol from 'typescript/lib/protocol';
+import tsp from 'typescript/lib/protocol';
```
Comment (rchl) [inline server/src/tsp-client.ts:13]: I'd suggest this instead:
```suggestion
import protocol, { CommandTypes } from 'typescript/lib/protocol';
```
Then you avoid all unnecessary changes and the diff is minimal.

Similarly in other files.

**Review by rchl — CHANGES_REQUESTED** [review]

File: server/src/protocol-translation.ts [hunk 3]
```diff
@@ -233,7 +238,15 @@ export function asTagsDocumentation(tags: tsp.JSDocTagInfo[]): string {
...
         case 'param': {
-            const body = (tag.text || '').split(/^([\w.]+)\s*-?\s*/);
+            // TODO: What does this actually look like? Is the original logic
+            // just doing something that the new type should do?
+            let formattedTag = '';
+            if (typeof tag.text === 'string') {
+                formattedTag = tag.text;
+            } else if (tag.text) {
+                formattedTag = tag.text.map(symbol => `${symbol.text} (${symbol.kind})`).join(', ');
+            }
```
Comment (rchl) [inline server/src/protocol-translation.ts:248]: I'd like something like:
```suggestion
const formattedTag = typeof tag.text === 'string' ? tag.text : asPlainText(tag.text);
```

and then update `asPlainText` to handle the `kind` in `SymbolDisplayPart`.

**Review by rchl — CHANGES_REQUESTED** [review]

File: server/src/protocol-translation.ts (hunk 3 above)
Comment (rchl) [reply server/src/protocol-translation.ts:248]: But to be fair, looking at the description of `kind` which is:

> The symbol's kind (such as 'className' or 'parameterName' or plain 'text')

I don't think we would necessarily want to to show the kind value. Certainly not when it's `text` but even in other cases it's probably not needed (at least for now).

So it seems fine to just do:

```
const formattedTag = typeof tag.text === 'string' ? tag.text : asPlainText(tag.text);
```
for now without updating `asPlainText`.

Comment (apexskier) [reply server/src/protocol-translation.ts:248]: Okay, I'm probably fine with that. I'm still playing with this, I want to see what it ends up looking like in https://github.com/apexskier/nova-typescript/ with full information.

Comment (rchl) [reply server/src/protocol-translation.ts:248]: Cool, please let me not (screenshots would be nice :)).

BTW. To enable typescript to send `SymbolDisplayPart` I guess some flag/setting needs to be enabled?

Comment (rchl) [reply server/src/protocol-translation.ts:248]: Though note the documentation that I've referenced above:

> The symbol's kind (such as 'className' or 'parameterName' or plain 'text')

Those kinds look too low-level to show in the UI anyway. If we would want to show those we probably would have to map them to UI-friendly strings.

Comment (apexskier) [reply server/src/protocol-translation.ts:248]: Looks like it's controlled by `UserPreferences.displayPartsForJSDoc`
https://github.com/microsoft/TypeScript/blob/62f9155a9d3c734854f3340e2b74fb799bf407fe/src/server/protocol.ts#L989

Comment (apexskier) [reply server/src/protocol-translation.ts:248]: At time of writing with the TODO experimental code, here's an example of what this looks like (obviously lots of work to be done)

TypeScript Extension – index d ts - Nova - 2021-07-02 at 21 34 45
TypeScript Extension – novaUtils ts - Nova - 2021-07-02 at 21 34 58

Comment (rchl) [reply server/src/protocol-translation.ts:248]: Right. So that `text` kind looks very redundant so far. I wonder in which cases we would get the other "kind"s.

File: server/src/protocol-translation.ts [hunk 4]
```diff
@@ -243,7 +255,7 @@ export function asTagDocumentation(tag: tsp.JSDocTagInfo): string {
                 }
                 return label + (doc.match(/\r\n|\n/g) ? '  \n' + doc : ` — ${doc}`);
             }
-            // fall-through
```
Comment (apexskier) [inline server/src/protocol-translation.ts:246]: There's no subsequent logic in the `switch` to fall through to.

File: server/src/protocol-translation.ts (hunk 3 above)
Comment (apexskier) [reply server/src/protocol-translation.ts]: FYI, I'm removing this in this initial PR. I'll do a secondary one to change the formatting behavior.

File: server/src/document-symbol.ts [hunk 5]
```diff
@@ -55,7 +55,7 @@ function collectDocumentSymbolsInRange(parent: tsp.NavigationTree, symbols: lsp.
     return shouldInclude;
 }
 
-export function collectSymbolInformations(uri: string, current: tsp.NavigationTree, symbols: lsp.SymbolInformation[], containerName?: string): boolean {
```
Comment (apexskier) [inline server/src/document-symbol.ts:58]: Typo fix

File: server/src/lsp-server.spec.ts [hunk 6]
```diff
@@ -491,106 +491,58 @@ describe('code actions', () => {
 
         assert.deepEqual(result, [
```
Comment (apexskier) [inline server/src/lsp-server.spec.ts:492]: TS 4 has different quick-actions. Note that this won't change if someone's running the language server on an older typescript version, since these are pulled from the tslib.

**Comment by rchl** [comment]
I've updated the lock file on master after previous changes (seems like that was forgotten before) so there is a conflict.

**Comment by rchl** [comment]
I hope you don't mind that I'm gonna push some minor changes to your branch.

File: server/src/lsp-server.ts [hunk 7]
```diff
@@ -487,7 +485,13 @@ export class LspServer {
             { language: 'typescript', value: result.body.displayString }
         ];
         const tags = asTagsDocumentation(result.body.tags);
-        contents.push(result.body.documentation + (tags ? '\n\n' + tags : ''));
+        let documentation: string;
+        if (typeof result.body.documentation === 'string') {
+            documentation = result.body.documentation;
+        } else {
+            documentation = result.body.documentation?.map(p => p.text).join('') || '';
```
Comment (rchl) [inline server/src/lsp-server.ts:492]: According to the types this can't be undefined so:
```suggestion
            documentation = result.body.documentation.map(p => p.text).join('') || '';
```

But can you instead use the `asPlainText` helper function everywhere instead?

File: server/src/protocol-translation.ts [hunk 8]
```diff
@@ -233,7 +211,13 @@ export function asTagsDocumentation(tags: tsp.JSDocTagInfo[]): string {
 export function asTagDocumentation(tag: tsp.JSDocTagInfo): string {
     switch (tag.name) {
         case 'param': {
-            const body = (tag.text || '').split(/^([\w.]+)\s*-?\s*/);
+            let text: string;
+            if (typeof tag.text === 'string') {
+                text = tag.text;
+            } else {
+                text = tag.text?.map(p => p.text).join('') || '';
```
Comment (rchl) [inline server/src/protocol-translation.ts:218]: Can use `asPlainText`?

File: server/src/protocol-translation.ts [hunk 9]
```diff
@@ -261,17 +245,24 @@ export function asTagBodyText(tag: tsp.JSDocTagInfo): string | undefined {
         return undefined;
     }
 
+    let text: string;
+    if (typeof tag.text === 'string') {
+        text = tag.text;
+    } else {
+        text = tag.text.map(p => p.text).join('');
```
Comment (rchl) [inline server/src/protocol-translation.ts:252]: Can use `asPlainText`?

**Review by rchl — CHANGES_REQUESTED** [review]

File: server/src/protocol-translation.ts [hunk 10]
```diff
@@ -233,7 +210,16 @@ export function asTagsDocumentation(tags: tsp.JSDocTagInfo[]): string {
...
-            const body = (tag.text || '').split(/^([\w.]+)\s*-?\s*/);
+            if (!tag.text) {
+                break;
+            }
+            let text: string;
+            if (typeof tag.text === 'string') {
+                text = tag.text;
+            } else {
+                text = asPlainText(tag.text);
+            }
```
Comment (rchl) [inline server/src/protocol-translation.ts:221]:
```suggestion
            const text = asPlainText(tag.text);
```

File: server/src/completion.ts [hunk 11]
```diff
@@ -203,11 +203,11 @@ export function asCodeActions(details: tsp.CompletionEntryDetails, filepath: str
...
 
-export function asDetail({ displayParts, source }: tsp.CompletionEntryDetails): string | undefined {
+function asDetail({ displayParts, sourceDisplay, source: deprecatedSource }: tsp.CompletionEntryDetails): string | undefined {
     const result: string[] = [];
-    const importPath = asPlainText(source);
-    if (importPath) {
-        result.push(`Auto import from '${importPath}'`);
+    const source = sourceDisplay || deprecatedSource;
+    if (source) {
+        result.push(`Auto import from '${asPlainText(source)}'`);
```
Comment (rchl) [inline server/src/completion.ts:210]: Any cases where this change has observable effect?

Comment (apexskier) [reply server/src/completion.ts:210]: I haven't found any, no. It should be backwards compatible with old versions of typescript since I'm still using `source`, but I didn't test explicitly.

Comment (rchl) [reply server/src/completion.ts:210]: Well, it looks like a good change anyway as `sourceDisplay` is supposed to be the human-readable one anyway.

**Review by rchl — APPROVED** [review]
Great stuff. Thanks.
